### PR TITLE
[Backport 2.x] Integrate Engine with decoupled Translog interfaces

### PR DESCRIPTION
**Pull request by satyajitg28**
### Description
The PR aims to Integrate Engine with decoupled Translog interfaces. Original CR [link](https://github.com/opensearch-project/OpenSearch/pull/3671). Ensuring `Engine` is backward compatible.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/engine/Engine.java [hunk 1]
```diff
@@ -842,6 +859,12 @@ public final CommitStats commitStats() {
      */
     public abstract long getPersistedLocalCheckpoint();
 
+    /**
+     * @return the latest checkpoint that has been processed but not necessarily persisted.
+     * Also see {@link #getPersistedLocalCheckpoint()}
+     */
+    public abstract long getProcessedLocalCheckpoint();
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/engine/Engine.java:866]: @satyajitg28  This is breaking one, we need to provide default implementation

File: server/src/main/java/org/opensearch/index/engine/Engine.java [hunk 2]
```diff
@@ -167,6 +171,8 @@ public final EngineConfig config() {
         return engineConfig;
     }
 
+    public abstract TranslogManager translogManager();
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/engine/Engine.java:174]: Same here, breaking one :(

**Review by reta — CHANGES_REQUESTED** [review]

**Comment by satyajitg28** [comment]
looks like a flaky test.
`./gradlew ':server:test' --tests "org.opensearch.index.shard.RemoveCorruptedShardDataCommandTests.testCorruptedTranslog"` passed on local.

File: test/framework/src/main/java/org/opensearch/index/engine/EngineTestCase.java [hunk 3]
```diff
@@ -328,24 +328,28 @@ public void tearDown() throws Exception {
...
+                replicaEngine.ensureOpen();
+                assertEngineCleanedUp(replicaEngine, replicaEngine.translogManager().getTranslog());
             }
         } finally {
             IOUtils.close(replicaEngine, storeReplica, engine, store, () -> terminate(threadPool));
         }
     }
 
+    protected void assertEngineCleanedUp(Engine engine, Translog translog) throws Exception {
+        if (engine.isClosed.get() == false) {
```
Comment (Bukhtawar) [inline test/framework/src/main/java/org/opensearch/index/engine/EngineTestCase.java:344]: Is this check redundant

File: server/src/test/java/org/opensearch/index/shard/IndexShardTests.java [hunk 4]
```diff
@@ -4036,8 +4036,7 @@ public void testCloseShardWhileResettingEngine() throws Exception {
         CountDownLatch closeDoneLatch = new CountDownLatch(1);
         IndexShard shard = newStartedShard(false, Settings.EMPTY, config -> new InternalEngine(config) {
             @Override
-            public InternalEngine recoverFromTranslog(TranslogRecoveryRunner translogRecoveryRunner, long recoverUpToSeqNo)
-                throws IOException {
+            public Engine recoverFromTranslog(TranslogRecoveryRunner translogRecoveryRunner, long recoverUpToSeqNo) throws IOException {
```
Comment (Bukhtawar) [inline server/src/test/java/org/opensearch/index/shard/IndexShardTests.java:4039]: Can we revert this change. Any reason why couldn't use `TranslogEventListener` here?

File: test/framework/src/main/java/org/opensearch/index/engine/EngineTestCase.java (hunk 3 above)
Comment (satyajitg28) [reply test/framework/src/main/java/org/opensearch/index/engine/EngineTestCase.java]: Yes, it was redundant. Removed it.

File: server/src/test/java/org/opensearch/index/shard/IndexShardTests.java (hunk 4 above)
Comment (satyajitg28) [reply server/src/test/java/org/opensearch/index/shard/IndexShardTests.java]: Reverted the above change. Using `TranslogEventListener`, the test is getting stuck. Will debug it further.

File: server/src/main/java/org/opensearch/index/engine/Engine.java (hunk 2 above)
Comment (satyajitg28) [reply server/src/main/java/org/opensearch/index/engine/Engine.java]: Fixed.

File: server/src/main/java/org/opensearch/index/engine/Engine.java (hunk 1 above)
Comment (satyajitg28) [reply server/src/main/java/org/opensearch/index/engine/Engine.java]: Fixed.

**Comment by satyajitg28** [comment]
Test failing: DiskThresholdDeciderIT.testHighWatermarkNotExceeded. Looks like a flaky test.

**Comment by satyajitg28** [comment]
Again this test is failing: DiskThresholdDeciderIT.testHighWatermarkNotExceeded.

**Comment by dblock** [comment]
@Bukhtawar care to take a look at this backport?

**Review by reta — APPROVED** [review]
